generator/geometry_holder.hpp

```diff
@@ -0,0 +1,249 @@
+
+  private:
+    Points const & m_src;
+    Points & m_dest;
+  };
+
+  void WriteOuterPoints(Points const & points, int i)
+  {
+    CHECK(m_geoFileGetter, ("m_geoFileGetter must be set to write outer points."));
+
+    // outer path can have 2 points in small scale levels
+    ASSERT_GREATER(points.size(), 1, ());
+
+    auto cp = m_header.GetCodingParams(i);
+
+    // Optimization: Store first point once in header for outer linear features.
+    cp.SetBasePoint(points[0]);
+    // Can optimize here, but ... Make copy of vector.
+    Points toSave(points.begin() + 1, points.end());
+
+    m_buffer.m_ptsMask |= (1 << i);
+    m_buffer.m_ptsOffset.push_back(feature::FeaturesCollector::GetFileSize(m_geoFileGetter(i)));
+    serial::SaveOuterPath(toSave, cp, m_geoFileGetter(i));
+  }
+
+  void WriteOuterTriangles(Polygons const & polys, int i)
+  {
+    CHECK(m_trgFileGetter, ("m_trgFileGetter must be set to write outer triangles."));
+
+    // tesselation
+    tesselator::TrianglesInfo info;
+    if (0 == tesselator::TesselateInterior(polys, info))
+    {
+      LOG(LINFO, ("NO TRIANGLES in", polys));
+      return;
+    }
+
+    auto const cp = m_header.GetCodingParams(i);
+
+    serial::TrianglesChainSaver saver(cp);
+
+    // points conversion
+    tesselator::PointsInfo points;
+    m2::PointU (*D2U)(m2::PointD const &, uint32_t) = &PointD2PointU;
+    info.GetPointsInfo(saver.GetBasePoint(), saver.GetMaxPoint(),
+                       std::bind(D2U, std::placeholders::_1, cp.GetCoordBits()), points);
+
+    // triangles processing (should be optimal)
+    info.ProcessPortions(points, saver, true);
+
+    // check triangles processing (to compare with optimal)
+    // serial::TrianglesChainSaver checkSaver(cp);
+    // info.ProcessPortions(points, checkSaver, false);
+
+    // CHECK_LESS_OR_EQUAL(saver.GetBufferSize(), checkSaver.GetBufferSize(), ());
+
+    // saving to file
+    m_buffer.m_trgMask |= (1 << i);
+    m_buffer.m_trgOffset.push_back(feature::FeaturesCollector::GetFileSize(m_trgFileGetter(i)));
+    saver.Save(m_trgFileGetter(i));
+  }
+
+  void FillInnerPointsMask(Points const & points, uint32_t scaleIndex)
+  {
+    auto const & src = m_buffer.m_innerPts;
+    CHECK(!src.empty(), ());
+
+    CHECK(feature::ArePointsEqual(src.front(), points.front()), ());
+    CHECK(feature::ArePointsEqual(src.back(), points.back()), ());
+
+    size_t j = 1;
+    for (size_t i = 1; i < points.size() - 1; ++i)
+    {
+      for (; j < src.size() - 1; ++j)
+      {
+        if (feature::ArePointsEqual(src[j], points[i]))
+        {
+          // set corresponding 2 bits for source point [j] to scaleIndex
+          uint32_t mask = 0x3;
+          m_buffer.m_ptsSimpMask &= ~(mask << (2 * (j - 1)));
+          m_buffer.m_ptsSimpMask |= (scaleIndex << (2 * (j - 1)));
+          break;
+        }
+      }
+
+      CHECK_LESS(j, src.size() - 1, ("Simplified point not found in source point array"));
```

Comment:
"Simplified point is not found in the source point array."

Reply:
done